### PR TITLE
Bazel build for opencensus bigtable integration

### LINKED ISSUE #958: Create an example program linking OpenCensus with Bazel
Create a program, with its own WORKSPACE that links OpenCensus and google-cloud-cpp.  In this version we should just link with Bazel, since OpenCensus does not support CMake yet (see https://github.com/census-instrumentation/opencensus-cpp/issues/86).

This depends on #957.

### PR DESCRIPTION
**Pull request by manish-qlogic**
This is an attempt to fix #958

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1230)

### CODE REVIEW

File: google/cloud/bigtable/examples/opencensus/WORKSPACE [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# Copyright 2018 Google LLC
```
Comment (coryan) [inline google/cloud/bigtable/examples/opencensus/WORKSPACE]: I think this file is ignored, does the example work without it?  If it does, please remove it.

File: google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+// limitations under the License.
+
+//! [all code]
+
+//! [bigtable includes]
+#include "google/cloud/bigtable/table.h"
+#include "google/cloud/bigtable/table_admin.h"
+//! [bigtable includes]
+
+#ifndef OPENCENSUS_DISABLED
```
Comment (coryan) [inline google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc]: Where is this macro defined?  It looks like it is not and this code only compiles if OpenCensus is available, please remove.

File: google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    }
+  }
+  //! [scan all]
+
+  // Delete the table
+  //! [delete table]
+  table_admin.DeleteTable(table_id);
+  //! [delete table]
+
+#ifndef OPENCENSUS_DISABLED
```
Comment (coryan) [inline google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc]: Is there a more deterministic way to terminate the reporting?  Maybe some OpenCensus API to query how many reports are pending?

File: google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    std::cerr << "Usage: " << cmd.substr(last_slash + 1)
+              << " <project_id> <instance_id> <table_id>" << std::endl;
+    return 1;
+  }
+
+  std::string const project_id = argv[1];
+  std::string const instance_id = argv[2];
+  std::string const table_id = argv[3];
+
+#ifndef OPENCENSUS_DISABLED
```
Comment (coryan) [inline google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc]: We need a README.md or a `*.dox` file that explains:

- How to setup stackdriver to read the open census data, just a link to the right documentation on the stackdriver and/or opencensus page would be enough.
- How to see the results of this example in the stack driver pages.

File: google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc (hunk 2 above)
Comment (manish-qlogic) [reply google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc]: removed.

File: google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc (hunk 3 above)
Comment (manish-qlogic) [reply google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc]: I couldn't figure out any API that could terminate gracefully ([code](https://github.com/census-instrumentation/opencensus-cpp)).  
I even checked [java example](https://opencensus.io/quickstart/java/tracing/) but couldn't find such API.

File: google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc (hunk 4 above)
Comment (manish-qlogic) [reply google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc]: included README.md

File: google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc]: Ack.

**Comment by manish-qlogic** [comment]
PTAL

File: google/cloud/bigtable/examples/opencensus/README.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
+# OpenCensus example for GCP Bigtable C++ Client.
+## Before you begin
```
[the remark this answers is not in the document]
Comment (manish-qlogic) [reply google/cloud/bigtable/examples/opencensus/README.md]: fixed.

File: google/cloud/bigtable/examples/opencensus/README.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+### Stats
+One can view stats captured using OpenCensus at Google Cloud Console as explained [here](https://github.com/census-instrumentation/opencensus-cpp/tree/master/examples/grpc#stats)
+
+### Traces
+One can view traces captured using OpenCensus at Google Cloud Console as explained [here](https://github.com/census-instrumentation/opencensus-cpp/tree/master/examples/grpc#tracing)
+
+## OpenCensus example
+### Build examp[le
+Binary `bigtable_opencensus` demonstrates the usage of OpenCensus for observability. This example current can be build using basel. One can build the binary using following command.
```
[the remark this answers is not in the document]
Comment (manish-qlogic) [reply google/cloud/bigtable/examples/opencensus/README.md]: fixed.

File: google/cloud/bigtable/examples/opencensus/README.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
+# OpenCensus example for GCP Bigtable C++ Client.
+## Before you begin
+### Setup StatckDriver
```
[the remark this answers is not in the document]
Comment (manish-qlogic) [reply google/cloud/bigtable/examples/opencensus/README.md]: fixed.

File: google/cloud/bigtable/examples/opencensus/README.md [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+Instructions to setup StackDriver can be found [here](https://github.com/census-instrumentation/opencensus-cpp/tree/master/examples/grpc#stackdriver)
```
[the remark this answers is not in the document]
Comment (manish-qlogic) [reply google/cloud/bigtable/examples/opencensus/README.md]: fixed.

File: google/cloud/bigtable/examples/opencensus/README.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+Binary `bigtable_opencensus` demonstrates the usage of OpenCensus for observability. This example current can be build using basel. One can build the binary using following command.
+
+```
+bazel build "//google/cloud/bigtable/...:all"
+```
+
+### Run example
+This example uses:
+ * The OpenCensus gRPC plugin to instrument the Bigtable gRPC calls.
+ * The Stackdriver exporter to export stats and traces.
```
[the remark this answers is not in the document]
Comment (manish-qlogic) [reply google/cloud/bigtable/examples/opencensus/README.md]: I think one example is enough, a good way to start using OpenCensus and Stackdriver with GCP Bigtable Client. We can add more examples later, we can also think of having examples on `Span` and `View`.

**Comment by manish-qlogic** [comment]
@sduskis 
PTAL

File: google/cloud/bigtable/examples/opencensus/README.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+The OpenCensus example will publish metrics and traces to [Stackdriver](https://cloud.google.com/stackdriver/), if you have not previously configured Stackdriver, follow the instructions in [...](https://github.com/census-instrumentation/opencensus-cpp/tree/master/examples/grpc#stackdriver) to enable Stackdriver for your Google Cloud project.
+
+### Stats
+One can view stats captured using OpenCensus at Google Cloud Console as explained [here](https://github.com/census-instrumentation/opencensus-cpp/tree/master/examples/grpc#stats)
+
+### Traces
+One can view traces captured using OpenCensus at Google Cloud Console as explained [here](https://github.com/census-instrumentation/opencensus-cpp/tree/master/examples/grpc#tracing)
+
+## OpenCensus example
+### Build examp[le
```
Comment (sduskis) [inline google/cloud/bigtable/examples/opencensus/README.md]: Please fix the typo `examp[le`

File: google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+  }
+
+  std::string const project_id = argv[1];
+  std::string const instance_id = argv[2];
+  std::string const table_id = argv[3];
+
+  // Register the OpenCensus gRPC plugin to enable stats and tracing in gRPC.
+  grpc::RegisterOpenCensusPlugin();
+
+  // Trace all outgoing RPCs.
```
Comment (sduskis) [inline google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc]: Please note that `ProbabilitySampler(1.0)` is not something that ought to be used for production, and is used for sample purposes only.  Ideally, there ought to be a link to what `ProbabilitySampler` does.

File: google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+  opencensus::trace::TraceConfig::SetCurrentTraceParams(
+      {128, 128, 128, 128, opencensus::trace::ProbabilitySampler(1.0)});
+
+  // For debugging, register exporters that just write to stdout.
+  opencensus::exporters::stats::StdoutExporter::Register();
+  opencensus::exporters::trace::StdoutExporter::Register();
+  opencensus::exporters::stats::StackdriverExporter::Register(
```
Comment (sduskis) [inline google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc]: Please add a comment about what a good value for `"bigtable-opencensus-0@unspecified-host"` ought to be.

File: google/cloud/bigtable/examples/opencensus/README.md (hunk 10 above)
Comment (manish-qlogic) [reply google/cloud/bigtable/examples/opencensus/README.md]: fixed.

File: google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc (hunk 11 above)
Comment (manish-qlogic) [reply google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc]: Added more comments and provided couple of reference to opencensus that talks about sampling.

File: google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc (hunk 12 above)
Comment (manish-qlogic) [reply google/cloud/bigtable/examples/opencensus/bigtable_opencensus.cc]: comment added.

**Comment by manish-qlogic** [comment]
@sduskis 
PTAL

**Comment by sduskis** [comment]
It looks great! @coryan, are you satisfied as well?

File: google/cloud/bigtable/examples/opencensus/README.md [hunk 13]
```diff
@@ -0,0 +1,35 @@
+# OpenCensus example for GCP Bigtable C++ Client.
+## Before you begin
+### About OpenCensus
+[OpenCensus](https://opencensus.io) is a vendor-agnostic single distribution of libraries to provide metrics collection and tracing for your services. In this example we show how one can integrate OpenCensus with the Cloud Bigtable C++ client library.
+
+### Setup Stackdriver
+The OpenCensus example will publish metrics and traces to [Stackdriver](https://cloud.google.com/stackdriver/), if you have not previously configured Stackdriver, follow the instructions in [...](https://github.com/census-instrumentation/opencensus-cpp/tree/master/examples/grpc#stackdriver) to enable Stackdriver for your Google Cloud project.
```
Comment (coryan) [inline google/cloud/bigtable/examples/opencensus/README.md]: One last nit, change that `follow the instructions in [...]` to maybe `follow [these instructions]`?

**Pull request merged**